File: ui/article-renderer/src/main/java/org/cru/godtools/article/activity/CategoriesActivity.java
package org.cru.godtools.article.activity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.MainThread;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentManager;

import org.cru.godtools.article.R;
import org.cru.godtools.article.fragment.CategoriesFragment;
import org.cru.godtools.base.tool.activity.BaseSingleToolActivity;

import java.util.Locale;

public class CategoriesActivity extends BaseSingleToolActivity {
    private static final String TAG_MAIN_FRAGMENT = "mainFragment";

    public static void start(@NonNull final Context context, @NonNull final String toolCode,
                             @NonNull final Locale language) {
        final Bundle extras = new Bundle(2);
        populateExtras(extras, toolCode, language);
        final Intent intent = new Intent(context, CategoriesActivity.class).putExtras(extras);
        context.startActivity(intent);
    }

    public CategoriesActivity() {
        super(false);
    }

    // region Lifecycle Events

    @Override
    protected void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_article_categories);
    }

    @Override
    protected void onStart() {
        super.onStart();
        loadInitialFragmentIfNeeded();
    }

    // endregion Lifecycle Events

    @MainThread
    private void loadInitialFragmentIfNeeded() {
        final FragmentManager fm = getSupportFragmentManager();

        // short-circuit if there is a currently attached fragment
        if (fm.findFragmentByTag(TAG_MAIN_FRAGMENT) != null) {
            return;
        }

        // update the displayed fragment
        assert mTool != null;
        getSupportFragmentManager().beginTransaction()
                .replace(R.id.frame, CategoriesFragment.newInstance(mTool, mLocale), TAG_MAIN_FRAGMENT)
                .commit();
    }
}

File: ui/article-renderer/src/main/java/org/cru/godtools/article/ui/categories/CategoriesActivity.kt
package org.cru.godtools.article.ui.categories

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.annotation.MainThread
import androidx.fragment.app.commit
import org.cru.godtools.article.R
import org.cru.godtools.article.ui.articles.startArticlesActivity
import org.cru.godtools.base.tool.activity.BaseArticleActivity
import org.cru.godtools.base.tool.activity.BaseSingleToolActivity
import org.cru.godtools.base.tool.analytics.model.SCREEN_CATEGORIES
import org.cru.godtools.base.tool.analytics.model.ToolAnalyticsScreenEvent
import org.cru.godtools.xml.model.Category
import java.util.Locale

fun Context.createCategoriesIntent(toolCode: String, language: Locale): Intent {
    return Intent(this, CategoriesActivity::class.java)
        .putExtras(BaseSingleToolActivity.buildExtras(this, toolCode, language))
}

fun Activity.startCategoriesActivity(toolCode: String, language: Locale) {
    startActivity(createCategoriesIntent(toolCode, language))
}

class CategoriesActivity : BaseArticleActivity(), CategorySelectedListener {
    // region Lifecycle
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (isFinishing) return
        setContentView(R.layout.activity_generic_tool_fragment)
        if (savedInstanceState == null) trackToolOpen(tool)
    }

    override fun onStart() {
        super.onStart()
        loadPrimaryFragmentIfNeeded()
    }

    override fun onResume() {
        super.onResume()
        eventBus.post(ToolAnalyticsScreenEvent(SCREEN_CATEGORIES, tool, locale))
    }

    override fun onCategorySelected(category: Category?) = startArticlesActivity(tool, locale, category?.id)
    // endregion Lifecycle

    @MainThread
    private fun loadPrimaryFragmentIfNeeded() {
        with(supportFragmentManager) {
            if (primaryNavigationFragment != null) return

            commit {
                val fragment = CategoriesFragment(tool, locale)
                replace(R.id.frame, fragment)
                setPrimaryNavigationFragment(fragment)
            }
        }
    }
}
